### ISSUE TITLE
Cache DLLs in DllImport

### ISSUE THREAD

**Issue by firox263** [gircore/gir.core#161]
Investigate whether we can/should cache the resolved DLL between calls to avoid a lookup each time we call a native method.

From #154:

> **na2axl:** Maybe a way to cache the loaded library, to avoid redundant loading ?

> **badcel:** I'm not sure if there is some caching on the side of NativeLibrary.TryLoad.
This reads a little bit like it: docs.
If you agreee I would avoid adding caching for now. If we see problems later or have definitive feedback that we are missing an opportunity here we can add the caching ourself.
gircore/coreteam what do you think?

> **badcel:** Hm I added a Console.WriteLine and the ImportResolver method is called several times. So it looks like it is not caching. But I'm not sure if we should add caching and if this could result in some unexpected results.
Edit: I found no search results which did some caching. Perhaps there are multiple contexts involved and multiple calls are correct?

> **na2axl:** I think we can check that by printing the returned IntPtr to the console, if the result is the same may be it's finally better to cache ?
But I've no problem to skip this for now.

**Comment by badcel** [comment]
I did some tests with the Quickstart project. I changed the `ImportResolver` method to look like this:

```C#
private static IntPtr ImportResolver(string libraryName, Assembly assembly, DllImportSearchPath? searchPath)
{
    var t = new System.Diagnostics.Stopwatch();
    t.Start();
            
    IntPtr libHandle = IntPtr.Zero;
    if (!NativeLibrary.TryLoad(DllName, assembly, DllImportSearchPath.LegacyBehavior, out libHandle))
    {
        libHandle = IntPtr.Zero;
    }
            
    t.Stop();
    System.Console.WriteLine("Loading: " + libraryName + " " + libHandle.ToString() + " " + t.Elapsed.ToString());

    return libHandle;
}
```

The output is:

```
Loading: Gtk 93879031024976 00:00:00.0120757
Loading: GObject 93879031052304 00:00:00.0002113
Loading: GObject 93879031052304 00:00:00.0001334
Loading: GObject 93879031052304 00:00:00.0001528
Loading: Gtk 93879031024976 00:00:00.0004266
Loading: Gtk 93879031024976 00:00:00.0001343
Loading: Gtk 93879031024976 00:00:00.0001310
Loading: Gtk 93879031024976 00:00:00.0001730
Loading: GObject 93879031052304 00:00:00.0001193
Instantiating Gtk.Window
Loading: GObject 93879031052304 00:00:00.0001325
Loading: GObject 93879031052304 00:00:00.0001519
Loading: GObject 93879031052304 00:00:00.0001223
Loading: GObject 93879031052304 00:00:00.0001938
Loading: GObject 93879031052304 00:00:00.0001851
Loading: GObject 93879031052304 00:00:00.0001183
Loading: GObject 93879031052304 00:00:00.0001259
Loading: GObject 93879031052304 00:00:00.0001188
Loading: GObject 93879031052304 00:00:00.0001284
Loading: Gtk 93879031024976 00:00:00.0001322
Instantiating Gtk.Notebook
Loading: Gtk 93879031024976 00:00:00.0001804
Loading: Gtk 93879031024976 00:00:00.0001340
Instantiating Gtk.Label
Instantiating Gtk.Label
Loading: Gtk 93879031024976 00:00:00.0003391
Loading: GObject 93879031052304 00:00:00.0001577
Loading: GObject 93879031052304 00:00:00.0001239
Loading: GObject 93879031052304 00:00:00.0001167
SwitchedPage: 0 with child Label
Loading: Gtk 93879031024976 00:00:00.0001900
Instantiating Gtk.Button
Instantiating Gtk.Label
SwitchedPage: 1 with child Button
Loading: GObject 93879031052304 00:00:00.0003175
Loading: Gtk 93879031024976 00:00:00.0001430
SwitchedPage: 0 with child Label
Loading: Gtk 93879031024976 00:00:00.0005501
SwitchedPage: 1 with child Button
Instantiating Gtk.Window
Instantiating Gtk.Button
Loading: Gtk 93879031024976 00:00:00.0001902
Instantiating Gtk.Window
Instantiating Gtk.Button
Instantiating Gtk.Window
Instantiating Gtk.Button
Instantiating Gtk.Window
Instantiating Gtk.Button
```

Some observations:
 - The first call to Gtk takes quite a bit longer than all subsequent calls
 - There are multiple calls for the same library
 - At the end I multiple times called native code (via opening the 2nd window from the Quickstart sample) and no further loading occured
 - The IntPtr values of the assemblies does not change

My conclusions:
- For me it looks like there is some caching happening as every call after the first one is quite a lot faster (more than factor 50).
- I'm still a bit worried because if there are repeated native calls like at the end there is no no call to load the dll. So I assume that there are some diffent contexts which are loaded and if the context does not change no call is executed.
- As we can't be sure where the DLL is located in the memory (caching it would be a pure guess) and it looks like there are somehow contexts involved I would not add caching until we get some definitive information which does not involve guessing.

But if we can get those information and it says that caching is fine, we can absolutely do it.

I made a small test which caches the Assembly information and it is quite a bit faster.

```
Loading: Gtk 94624075916464 00:00:00.0115270
Loading: GObject 94624075940896 00:00:00.0002419
Loading: GObject 94624075940896 00:00:00.0000208
Loading: GObject 94624075940896 00:00:00.0000011
Loading: Gtk 94624075916464 00:00:00.0000014
Loading: Gtk 94624075916464 00:00:00.0000007
Loading: Gtk 94624075916464 00:00:00.0000006
Loading: Gtk 94624075916464 00:00:00.0000022
Loading: GObject 94624075940896 00:00:00.0000009
Instantiating Gtk.Window
Loading: GObject 94624075940896 00:00:00.0000017
Loading: GObject 94624075940896 00:00:00.0000014
Loading: GObject 94624075940896 00:00:00.0000008
Loading: GObject 94624075940896 00:00:00.0000022
Loading: GObject 94624075940896 00:00:00.0000019
Loading: GObject 94624075940896 00:00:00.0000006
Loading: GObject 94624075940896 00:00:00.0000007
Loading: GObject 94624075940896 00:00:00.0000008
Loading: GObject 94624075940896 00:00:00.0000008
Loading: Gtk 94624075916464 00:00:00.0000011
Instantiating Gtk.Notebook
Loading: Gtk 94624075916464 00:00:00.0000012
Loading: Gtk 94624075916464 00:00:00.0000006
Instantiating Gtk.Label
Instantiating Gtk.Label
Loading: Gtk 94624075916464 00:00:00.0000014
Loading: GObject 94624075940896 00:00:00.0000013
Loading: GObject 94624075940896 00:00:00.0000007
Loading: GObject 94624075940896 00:00:00.0000006
SwitchedPage: 0 with child Label
Loading: Gtk 94624075916464 00:00:00.0000009
Instantiating Gtk.Button
Instantiating Gtk.Label
SwitchedPage: 1 with child Button
Loading: GObject 94624075940896 00:00:00.0000014
Loading: Gtk 94624075916464 00:00:00.0000008
SwitchedPage: 0 with child Label
Loading: Gtk 94624075916464 00:00:00.0000028
SwitchedPage: 1 with child Button
Instantiating Gtk.Window
Instantiating Gtk.Button
Loading: Gtk 94624075916464 00:00:00.0000034
Instantiating Gtk.Window
Instantiating Gtk.Button
Instantiating Gtk.Window
Instantiating Gtk.Button
Instantiating Gtk.Window
Instantiating Gtk.Button
Instantiating Gtk.Window
Instantiating Gtk.Button
Instantiating Gtk.Window
Instantiating Gtk.Button
```

The code looked like:

```C#
private static IntPtr ImportResolver(string libraryName, Assembly assembly, DllImportSearchPath? searchPath)
{
    var t = new System.Diagnostics.Stopwatch();
    t.Start();
    IntPtr libHandle = IntPtr.Zero;

    if (cache.TryGetValue(libraryName, out libHandle))
    {
        t.Stop();
    }
    else
    {
        if (!NativeLibrary.TryLoad(DllName, assembly, DllImportSearchPath.LegacyBehavior, out libHandle))
        {
            libHandle = IntPtr.Zero;
        }
        else
        {
            cache[libraryName] = libHandle;
        }
        
        t.Stop();
    }

    System.Console.WriteLine("Loading: " + libraryName + " " + libHandle.ToString() + " " + t.Elapsed.ToString());

    return libHandle;
}
```

**Comment by na2axl** [comment]
Definitely good work @firox263 🚀

For the native dll loading, AFAIK from my experience while building our game engine I can say that the DLL loader cache the returned IntPtr for a specific native function. So if the loader load a DLL firstly for the function A, the second time that A will be called the cached pointer of A (and the cached one of its DLL) will be returned, but if the loader loads the function B the loader will resolve again the same DLL, but will go faster because he already know the path and some other info of that DLL, and then will load the DLL again to finally return a pointer for the function B.

We have ended to suppose this behavior while building bindings for OpenGL, and we finally decided to cache the returned pointer of the DLL and even for each native functions, and no problems has been raised after (and theses OpenGL functions are called several times on each frames of a game).

I think is also the reason why after several calls the loader is not running because the loaded function are reused.

Maybe we can give a try to caching and do several tests with it, if it's not satisfying we can always go back to the normal behavior.
